File: ListaDeAfazeres.Server/Modules/Features/ToDoTask/Model/ToDoTaskModel.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using ListaDeAfazeres.Server.Modules.Features.ToDoTask.DTOs;
using ListaDeAfazeres.Server.Modules.Utils.Model;
using TypeGen.Core.TypeAnnotations;

namespace ListaDeAfazeres.Server.Modules.Features.ToDoTask.Model
{
    [ExportTsClass]
    public class ToDoTaskModel : BaseModel
    {
        public ToDoTaskModel() {}

        public ToDoTaskModel(ToDoTaskCreateDTO dto)
        {
            Title = dto.Title;
            MarkToDoTask(dto.isCompleted);
        }

        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; private set; }

        required public string Title { get; set; }

        public DateTime CreatedAt { get; private set; } = DateTime.UtcNow;

        public DateTime? CompletedAt { get; set; }

        public void MarkToDoTask(bool isCompleted)
        {
            CompletedAt = isCompleted ? DateTime.UtcNow : null;
        }
    }
}

File: listadeafazeres.client/src/models/ToDoTask/ToDoTaskModel.ts
export class ToDoTaskModel {
  constructor(
    public id: number,
    public title: string,
    public createdAt: Date,
    public completedAt: Date | null,
  ) {}
}
export function toDoTaskModelFactory(data: any): ToDoTaskModel {
  return new ToDoTaskModel(
    data.id,
    data.title,
    new Date(data.createdAt),
    data.completedAt ? new Date(data.completedAt) : null,
  );
}
export class ToDoTaskDTO {
  constructor(
    public title: string,
    public isCompleted: boolean,
  ) {}
}
